source/opt/value_number_table.cpp

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "value_number_table.h"
+
+#include <algorithm>
+
+namespace spvtools {
+namespace opt {
+
+uint32_t ValueNumberTable::GetValueNumber(spvtools::ir::Instruction* inst) {
+  // TODO: Need to implement the substitution of operands by their value number
+  // before hashing.
+  // TODO: Implement a normal form for opcodes that commute like integer
+  // addition.  This will let us know that a+b is the same value as b+a.
+  assert(inst->result_id() != 0 &&
+         "inst must have a result id to get a value number.");
+
+  // Check if this instruction already has a value.
+  auto id_to_val = id_to_value_.find(inst->result_id());
+  if (id_to_val != id_to_value_.end()) {
+    return id_to_val->second;
+  }
+
+  // If it is a load from memory that can be modified, we have to assume the
+  // memory has been modified, so we give it a new value number.
+  //
+  // Note that this test will also handle volatile loads because they are not
+  // read only.  However, if this is ever relaxed because we analyze stores, we
+  // will have to add a new case for volatile loads.
+  if (inst->IsLoad() && !inst->IsReadOnlyLoad()) {
+    uint32_t value_number = TakeNextValueNumber();
+    id_to_value_[inst->result_id()] = value_number;
+    return value_number;
+  }
+
+  // Functions calls that modify or use external memory may return a different
+  // value even if the operands are the same.  Unless we can prove the function
+  // is pure, we have to assume each call returns a different value.
+  if (inst->opcode() == SpvOpFunctionCall) {
+    uint32_t value_number = TakeNextValueNumber();
+    id_to_value_[inst->result_id()] = value_number;
+    return value_number;
+  }
+
+  // Atomic operation often represent shared memory, so something else could
+  // have changed the inputs.  It atomic instruction need a unique value number.
+  if (inst->IsAtomicOp()) {
+    uint32_t value_number = TakeNextValueNumber();
+    id_to_value_[inst->result_id()] = value_number;
+    return value_number;
+  }
+
```

Comment:
Should grant new value numbers for all group instructions since they represent communication with the outside world.

Similar for pipe reads : OpPipeRead, OpReservedReadPipe.

... which makes me think that you should be using a whitelist of what instructions are actually combinators.  
See IsCombinator*  aggressive_dead_code_elim_pass.

The code in the rest of this function only correctly says yes when the instructions are combinators.

Reply:
Derivative instructions are another case where there's communication with other invocations behind your back.  Treat them like a group instruction, or a volatile load for example.

Reply:
Done, I factored out the combinator stuff from ADCE into the IRContext.  Now it is generally available.